### PR TITLE
Add nextEventualValueOrRedactionFor to MockKafkaProducer to allow testing for redactions.

### PR DESCRIPTION
**Pull request by adriaanm**
`nextEventualValueOrRedactionFor` returns the next message for the given topic and key wrapped in Some, or None if a redaction was received.

### CODE REVIEW

File: src/main/scala/io/github/jchapuis/fs2/kafka/mock/impl/NativeMockKafkaProducer.scala [hunk 1]
```diff
@@ -76,6 +76,18 @@ private[mock] class NativeMockKafkaProducer(
       else IO.pure(false)
   ).map { case (_, value) => value }
 
+  def nextEventualValueOrRedactionFor[K: Eq, V](topic: String, key: K)(implicit
```
Comment (jchapuis) [inline src/main/scala/io/github/jchapuis/fs2/kafka/mock/impl/NativeMockKafkaProducer.scala:79]: i notice usage of the term "Redaction" here vs. "Redacted" elsewhere, is it intentional?

File: src/main/scala/io/github/jchapuis/fs2/kafka/mock/impl/NativeMockKafkaProducer.scala [hunk 2]
```diff
@@ -106,24 +137,26 @@ private[mock] class NativeMockKafkaProducer(
...
       for {
         isSelected <- recordSelector(record)
         key <-
           if (isSelected) keyDeserializer.deserialize(topic, Headers.empty, record.key).map(Option(_))
           else IO(None)
         value <-
-          if (isSelected) valueDeserializer.deserialize(topic, Headers.empty, record.value).map(Option(_))
-          else IO(None)
+          (if (isSelected) {
+             if (record.value eq null) IO(Some(None))
```
Comment (jchapuis) [inline src/main/scala/io/github/jchapuis/fs2/kafka/mock/impl/NativeMockKafkaProducer.scala:157]: wondering if using `Option(value)` would have done the trick also?

File: src/main/scala/io/github/jchapuis/fs2/kafka/mock/impl/NativeMockKafkaProducer.scala (hunk 1 above)
Comment (adriaanm) [reply src/main/scala/io/github/jchapuis/fs2/kafka/mock/impl/NativeMockKafkaProducer.scala:79]: yeah, I was thinking "produced" / "a production" == "redacted" / "a redaction" (== "evaluated" / "a value")

File: src/main/scala/io/github/jchapuis/fs2/kafka/mock/impl/NativeMockKafkaProducer.scala (hunk 2 above)
Comment (adriaanm) [reply src/main/scala/io/github/jchapuis/fs2/kafka/mock/impl/NativeMockKafkaProducer.scala:157]: Yup, I tried that first, but it was kind of tricky to the point that I felt it would be hard to follow, so I decided to make it explicit (i.e. I gave up ;-))

Comment (adriaanm) [reply src/main/scala/io/github/jchapuis/fs2/kafka/mock/impl/NativeMockKafkaProducer.scala:157]: Consider me nerdsniped:
```
if (isSelected) Option(record.value).traverse(valueDeserializer.deserialize(topic, Headers.empty, _)).map(Some(_))
else IO(None)
```

I think this is equivalent, but I had to think a lot harder :-) Up to you what you prefer.

Comment (jchapuis) [reply src/main/scala/io/github/jchapuis/fs2/kafka/mock/impl/NativeMockKafkaProducer.scala:157]: glad i could get you to rev up the engine 😄  ⚙️ I like the second one better I think, nested ifs and 'null' kind of make me restless nowadays 😆

**Review by jchapuis — APPROVED** [review]

File: src/main/scala/io/github/jchapuis/fs2/kafka/mock/impl/NativeMockKafkaProducer.scala [hunk 3]
```diff
@@ -146,20 +145,17 @@ private[mock] class NativeMockKafkaProducer(
...
-          else IO(None)
-        value <-
-          (if (isSelected) {
-             if (record.value eq null) IO(Some(None))
-             else valueDeserializer.deserialize(topic, Headers.empty, record.value).map(v => Some(Some(v)))
-           } else IO(None)): IO[Option[Option[V]]]
-      } yield key.zip(value).map { case (k, v) => (index, k, v) }
+    .flatTraverse { case (record, index) =>
+      recordSelector(record)
+        .ifM(
```
Comment (adriaanm) [inline src/main/scala/io/github/jchapuis/fs2/kafka/mock/impl/NativeMockKafkaProducer.scala:150]: I tried hard to get rid of this if and somehow lift the condition to Some/None or something more structure, but couldn't figure it out.

**Review by jchapuis — DISMISSED** [review]

**Review by jchapuis — APPROVED** [review]
